Musicca/app/src/main/java/com/example/musicca/adapters/QueueAdapter.java

```diff
@@ -100,23 +101,108 @@ protected void publishResults(CharSequence charSequence, FilterResults filterRes
+
+                        Log.d("FILTER", "all fields not empty");
+                        filteredList.add(song);
+                    }
+                    // else if first two fields are filled, last one is empty
+                } else if (!TextUtils.isEmpty(year) && !TextUtils.isEmpty(songTitle) && TextUtils.isEmpty(songArtist)){
+                    if (song.getYear().toLowerCase().contains(year) &&
+                            song.getTitle().toLowerCase().contains(songTitle)) {
+
+                        Log.d("FILTER", "two fields not empty");
+                        filteredList.add(song);
+                    }
+                    // else if first if filled, second empty, third filled
+                } else if (!TextUtils.isEmpty(year) && TextUtils.isEmpty(songTitle) && !TextUtils.isEmpty(songArtist)){
+                    if (song.getYear().toLowerCase().contains(year) &&
+                            song.getArtist().toLowerCase().contains(songArtist)) {
+
+                        Log.d("FILTER", "two fields not empty");
+                        filteredList.add(song);
+                    }
+                    // else if last two are filled, first one is empty
+                } else if (TextUtils.isEmpty(year) && !TextUtils.isEmpty(songTitle) && !TextUtils.isEmpty(songArtist)){
+                    if (song.getTitle().toLowerCase().contains(songTitle) &&
+                            song.getArtist().toLowerCase().contains(songArtist)) {
+
+                        Log.d("FILTER", "two fields not empty");
+                        filteredList.add(song);
+                    }
+                    // else if first one is filled, the rest are empty
+                } else if (!TextUtils.isEmpty(year) && TextUtils.isEmpty(songTitle) && TextUtils.isEmpty(songArtist)){
+                    if (song.getYear().toLowerCase().contains(year)) {
+
+                        Log.d("FILTER", "one field not empty");
+                        filteredList.add(song);
+                    }
+                    // else if second one is filled, the rest are empty
+                } else if (TextUtils.isEmpty(year) && !TextUtils.isEmpty(songTitle) && TextUtils.isEmpty(songArtist)){
+                    if (song.getTitle().toLowerCase().contains(songTitle)) {
+
+                        Log.d("FILTER", "one field not empty");
+                        filteredList.add(song);
+                    }
+                    // else if third one is filled , the rest are empty
+                } else if (TextUtils.isEmpty(year) && TextUtils.isEmpty(songTitle) && !TextUtils.isEmpty(songArtist)){
+                    Log.d("FILTER", "one field not empty1");
+                    if (song.getArtist().toLowerCase().contains(songArtist)) {
+
+                        Log.d("FILTER", "one field not empty2");
+                        filteredList.add(song);
+                    }
+                }
```

Comment:
I think you can simplify the code here. With the current approach of enumerating all the combinations of whether each field is empty, it is error-prone and it will be hard for you to add another field to another filter.
Basically you want to skip checking a field if the field is empty right? How about something like this (in pseudocode):
Say I have two fields from my filter: field1, field2, and the following logic checks if an entry matches the filter.
```
boolean selected = true;
if (field1 is not empty && field1 does not match) {
   selected = false;
}
if (field2 is not empty && field2 does not match) {
   selected = false;
}
if (selected == true) {
  filteredList.add(song);
}
```
Simplifying this even more:
```
if ((field1 is empty || field1 matches) && (field2 is empty || field2 matches)) {
  filteredList.add(song);
}
```